lib/src/protocol/kawa_h1/mod.rs

```diff
@@ -0,0 +1,1933 @@
+                error!("PROXY session {:?} back error", self.frontend_token);
+                trace!("backend_hup: {:?}", state_result);
+                match state_result {
+                    StateResult::Continue => {}
+                    StateResult::CloseBackend => self.close_backend(proxy.clone(), metrics),
+                    StateResult::CloseSession => return SessionResult::Close,
+                    StateResult::ConnectBackend | StateResult::Upgrade => unreachable!(),
+                }
+            }
+
+            counter += 1;
+        }
+
+        if counter == max_loop_iterations {
+            error!(
+                "PROXY\thandling session {:?} went through {} iterations, there's a probable infinite loop bug, closing the connection",
+                self.frontend_token, max_loop_iterations
+            );
+            incr!("http.infinite_loop.error");
+
+            self.print_state(&format!("{:?}", self.context.protocol));
+
+            return SessionResult::Close;
+        }
+
+        SessionResult::Continue
+    }
+
+    pub fn timeout_status(&self) -> TimeoutStatus {
+        if self.request_stream.is_main_phase() {
+            if self.response_stream.is_initial() {
+                TimeoutStatus::WaitingForResponse
+            } else {
+                TimeoutStatus::Response
+            }
+        } else {
+            if self.keepalive_count > 0 {
+                TimeoutStatus::WaitingForNewRequest
+            } else {
+                TimeoutStatus::Request
+            }
+        }
+    }
+}
+
+impl<Front: SocketHandler, L: ListenerHandler + L7ListenerHandler> SessionState for Http<Front, L> {
+    fn ready(
+        &mut self,
+        session: Rc<RefCell<dyn crate::ProxySession>>,
+        proxy: Rc<RefCell<dyn L7Proxy>>,
+        metrics: &mut SessionMetrics,
+    ) -> SessionResult {
+        self.ready_inner(session, proxy, metrics)
+    }
+
+    fn update_readiness(&mut self, token: Token, events: Ready) {
+        if self.frontend_token == token {
+            self.frontend_readiness.event |= events;
+        } else if self.backend_token == Some(token) {
+            self.backend_readiness.event |= events;
+        }
+    }
+
+    // TODO: State should take responsibility for closing sockets and token entries?
+    // how about no
+    fn close(&mut self, proxy: Rc<RefCell<dyn L7Proxy>>, metrics: &mut SessionMetrics) {
+        self.close_backend(proxy, metrics);
+
+        //if the state was initial, the connection was already reset
+        if !self.request_stream.is_initial() {
+            gauge_add!("http.active_requests", -1);
+
+            if let Some(b) = self.backend.as_mut() {
+                let mut backend = b.borrow_mut();
+                backend.active_requests = backend.active_requests.saturating_sub(1);
+            }
+        }
+    }
+
+    fn timeout(&mut self, token: Token, metrics: &mut SessionMetrics) -> StateResult {
```

Comment:
please rename the function to an explicit name. What does it do? `check_timeouts`? `Set_timeouts` ?